### PR TITLE
Add window func related logic plan to proto ability.

### LINKED ISSUE #4480: Add window func related logic plan to proto ability.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ted-Jiang**
Signed-off-by: yangjiang <yangjiang@ebay.com>

# Which issue does this PR close?

Closes #4480.

# Rationale for this change

Help support window function in ballista.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/proto/src/from_proto.rs [hunk 1]
```diff
@@ -806,11 +806,20 @@ pub fn parse_expr(
...
+                        Some(x) => {
+                            vec![x]
+                        }
+                    };
+
                     Ok(Expr::WindowFunction {
                         fun: datafusion_expr::window_function::WindowFunction::BuiltInWindowFunction(
                             built_in_function,
                         ),
-                        args: vec![parse_required_expr(&expr.expr, registry, "expr")?],
```
Comment (Ted-Jiang) [inline datafusion/proto/src/from_proto.rs:813]: Unlike  WindowFunction  with `AggrFunction`  like `min(col)`. `max(col)`. `sum(col)` it must have required_exp

Built-in WindowFunction like `ROW_NUMBER()` . `RANK()`. `DENSE()`   have to have this value

**Comment by Ted-Jiang** [comment]
@andygrove PATL , seems this part not have a test, have tested with ballista.

File: datafusion/proto/src/from_proto.rs [hunk 2]
```diff
@@ -806,11 +806,20 @@ pub fn parse_expr(
                         .ok_or_else(|| Error::unknown("BuiltInWindowFunction", *i))?
                         .into();
 
+                    let args = match parse_optional_expr(&expr.expr, registry)? {
+                        None => {
+                            vec![]
+                        }
+                        Some(x) => {
```
Comment (Ted-Jiang) [inline datafusion/proto/src/from_proto.rs:813]: is there a better approach  ? 🤔

Comment (alamb) [reply datafusion/proto/src/from_proto.rs:813]: How about something like (untested):

```rust
let args = parse_optional_expr(&expr.expr, registry)?
  .map(Vec::new)
  .unwrap_or_else(|| vec![]);
```

Not sure if that is better but it is more functional

**Review by alamb — COMMENTED** [review]
Code looks good to me @Ted-Jiang 

can you please add a test to ensure we don't introduce regressions in the future?

File: datafusion/proto/src/from_proto.rs (hunk 2 above)
Comment (Ted-Jiang) [reply datafusion/proto/src/from_proto.rs:813]: souds good, i try it but seems `Vec::new` can not create one size vec with one agrs.

**Comment by Ted-Jiang** [comment]
> Code looks good to me @Ted-Jiang
> 
> can you please add a test to ensure we don't introduce regressions in the future?

@alamb Thanks for your advice, add test in [add test.](https://github.com/apache/arrow-datafusion/pull/4485/commits/b2be1bd333cce0baed4482d13319668141a72d33)

File: datafusion/proto/src/lib.rs [hunk 3]
```diff
@@ -1331,4 +1333,67 @@ mod roundtrip_tests {
...
+            partition_by: vec![col("col1")],
+            order_by: vec![col("col2")],
+            window_frame: Some(WindowFrame::default()),
+        };
+
+        // 3. with window_frame with row numbers
+        let range_number_frame = WindowFrame {
+            units: WindowFrameUnits::Range,
+            start_bound: WindowFrameBound::Preceding(ScalarValue::UInt64(Some(2))),
+            end_bound: WindowFrameBound::Following(ScalarValue::UInt64(Some(2))),
```
Comment (Ted-Jiang) [inline datafusion/proto/src/lib.rs:1366]: without this commit, this test will fail.

File: datafusion/proto/src/from_proto.rs (hunk 2 above)
Comment (alamb) [reply datafusion/proto/src/from_proto.rs:813]: Maybe 

```rust
let args = parse_optional_expr(&expr.expr, registry)?
  .map(|e| vec![e])
  .unwrap_or_else(|| vec![]);
```

Comment (Ted-Jiang) [reply datafusion/proto/src/from_proto.rs]: ye 😂,  thanks

**Pull request merged**